### ISSUE TITLE
Vertex shader for Polygon2D CanvasItem material transforms unexpectedly with batching enabled

### ISSUE THREAD

**Issue by TidB** [godotengine/godot#47090]
**Godot version:**
Godot 3.2.4.rc5 - was working fine with 3.2.4.rc4, 3.2.4.rc3 and 3.2.3

**OS/device including version:**
Windows 10 Home 20H2
GPU: RX 580, Radeon Software 20.12.1, GLES3

**Issue description:**
Using a CanvasItem vertex shader that uses the vertex input and output on a Polygon2D produces unexpected results when batching is enabled. The most minimal example has the output vertex having the same value as the input vertex (`VERTEX = VERTEX`). The expected behavior would be a Polygon2D behaving the exact same as one without a shader would. The resulting shape seems to be affected by the viewport, where the rendered shape and the actual location of the node only overlap when the viewport is adjusted manually to have both at position (0,0). As such, other transformations behave unexpectedly as well (moving, zooming and scaling in the editor):
![vertex](https://user-images.githubusercontent.com/6843686/111453365-cbe76500-8713-11eb-9434-c5695f430cea.gif)
The issue appears both in the editor and in the game. Disabling batching makes the issue disappear. This issue does not appear when applying the same shader to a Sprite.

This seems to be related to #46897 and the subsequent PR #46901, which also mentions possible regressions. In contrast, this issue does not use GDScript at all and seems to be limited to Polygon2D.

**Steps to reproduce:**
1. Create a Polygon2D node and create a shape with it
2. Add a new ShaderMaterial with a new Shader to the Polygon2D with the content:
```
shader_type canvas_item;

void vertex() {
	VERTEX = VERTEX;
}
```

Alternatively, a VisualShader item can be created with just the `vertex` input being connected to the `Vertex` shader output.

**Minimal reproduction project:**
[vertex-test.zip](https://github.com/godotengine/godot/files/6155537/vertex-test.zip)

The project contains `PolygonShader` (a Polygon2D with the aforementioned vertex shader as text), `PolygonVisualShader` (a different Polygon2D with the equivalent VisualShader), `SpriteShader` (a Sprite with the same vertex shader as text) and `MovementTest` (a playable scene showing the movement of a normal Polygon without a shader vs the SpriteShader vs the PolygonVisualShader).

**Comment by lawnjelly** [comment]
This was quite simple to fix, it was caused by applying the transform twice. Any time you read from (local) VERTEX in a custom shader, it has to use `large_fvf` which sends the transform as an attribute rather than a uniform, and applies the transform after the custom shader.

This means that there shouldn't be any CPU transform. This is already done for rects but not for polys or lines. I've made a fix.

There's just one further case which cannot be supported, which is reading VERTEX in a software skinned poly. If this is done, you would have to either switch off software transform in project settings, or maybe I can automatically detect this and revert to legacy. I'm trying the auto detect now, if it proves too difficult I can separate these into 2 PRs, as the latter should be a rare case and is fixable by changing a project setting.
